runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -382,21 +380,35 @@ function TSHighlighter._on_spell_nav(_, _, buf, srow, _, erow, _)
 end
 
 ---@private
----@param _win integer
 ---@param buf integer
 ---@param topline integer
 ---@param botline integer
-function TSHighlighter._on_win(_, _win, buf, topline, botline)
+function TSHighlighter._on_win(_, _, buf, topline, botline)
   local self = TSHighlighter.active[buf]
-  if not self then
+  if not self or self.parsing then
+    return false
+  end
+  self.parsing = self.tree:parse({ topline, botline + 1 }, function(trees)
+    if trees then
+      self:_async_parse_callback(buf)
+    end
+  end) --[[@as boolean]]
+  if self.parsing then
     return false
   end
-  self.tree:parse({ topline, botline + 1 })
-  self:prepare_highlight_states(topline, botline + 1)
   self.redraw_count = self.redraw_count + 1
+  self:prepare_highlight_states(topline, botline)
   return true
 end
 
+---@param buf integer
+function TSHighlighter:_async_parse_callback(buf)
+  if self.parsing then
+    self.parsing = false
+    api.nvim__redraw({ buf = buf, valid = false, flush = true })
```

Comment:
Interestingly, this `nvim__redraw()` seems to be redundant, seemingly all we need to do in this callback is to unset `self.parsing`. (Currently) the `on_changedtree` callback already triggers a sufficient redraw during parsing. The code responsible has a TODO note attached to it though so maybe that will change:
https://github.com/neovim/neovim/blob/c51bf5a6b24928ac04d0bb129b1b424d4c78f28d/runtime/lua/vim/treesitter/languagetree.lua#L623-L626.

In which case maybe we should keep this anyhow, not sure about the arguments though.
If other redraws are likely, `flush`ing here is actually unwanted so maybe should be set to `false`.

I used `valid = false` as opposed to a `topline/botline` range which we could still pass here, because there there is no guarantee that the topline will still be the same. IDK if this makes a difference in practice, (currently it doesn't because of the above🤷🏻).

Reply:
Interesting... have you noticed this for large files as well? For me removing the redraw is the same for small files that can presumably get parsed synchronously, but with large files my highlights will not apply without the redraw

Reply:
Ah you're right, I was testing in a large fall but small window😅
In that case, I think this is the best we can do:
```suggestion
    api.nvim__redraw({ buf = buf, valid = false, flush = false })
```

Unless if we keep track of what lines are actually drawn since the `on_win` call associated with the async callback. Then we could limit the range to those lines...

Reply:
Perfect, I'll update it :+1:  Out of curiosity, what is the difference here between `buf = buf` and `win = win`, or `valid = false` and `range = ...`?

Reply:
End line in `nvim__redraw()` overflows:
```console
Changes: { { 0, 0, 0, 4294967295, 4294967295, 4294967295 } }
Redrawing from 1 to 0  
```

Also it might not work because when the ranges change, both old and new need to be redrawn. And the callback redraws only one set of ranges.

Reply:
`buf = buf` marks all windows that contain a buffer to be redrawn, `win = win` targets a specific window, `valid = false` redraws from topline until botline and `range` redraws a specific range.

Since the user can still scroll during the asynchronous parse, when it returns, the topline may not be the same as the one we might have passed to this callback from `on_win`. What we could do is keep track of the min/max line number redrawn since a particular async parse started. That could potentially be more efficient (would have to still return `true` in `on_win`, then update the `min/max` in `on_line` but return without placing marks while it is parsing...).